dsgrid/query/models.py

```diff
@@ -0,0 +1,305 @@
+import abc
+from pathlib import Path
+from typing import Any, List, Optional, Union
+
+import pyspark.sql.functions as F
+from pydantic import Field, root_validator, validator
+
+from dsgrid.data_models import DSGBaseModel
+
+# from dsgrid.dataset.dimension_filters import (
+#     DimensionFilterValueModel,
+#     DimensionFilterExpressionModel,
+#     DimensionFilterExpressionRawModel,
+#     DimensionFilterColumnOperatorModel,
+# )
+from dsgrid.dimension.base_models import DimensionType
+from dsgrid.utils.files import compute_hash, load_data
+
+
+class FilteredDatasetModel(DSGBaseModel):
+
+    dataset_id: str = Field(title="dataset_id", description="Dataset ID")
+    filters: List[Any] = Field(
+        # Union[
+        #     DimensionFilterValueModel,
+        #     DimensionFilterExpressionModel,
+        #     DimensionFilterExpressionRawModel,
+        #     DimensionFilterColumnOperatorModel,
+        # ]
+        # ] = Field(
+        title="filters",
+        description="Dimension filters to apply to the dataset'",
+    )
+
+
+class MetricReductionModel(DSGBaseModel):
+    """Reduces the metric values for each set of unique dimension combinations, excluding
+    query_name."""
```

Comment:
I was trying to express the concept as generically as possible, such that it would apply to a data format with pivoted columns as well as a long, narrow format. Also, with or without time. For each combination of dimension (excluding the one referred to by query_name) there will be one or more metric values that can be reduced with an operation. Does that make sense? If so, how would you write it?:

Reply:
This concept will go away. This PR needs to combine `MetricReductionModel` and `AggregationModel`. The code needs to detect whether a groupBy column requires row-wise or column-wise aggregation and do the right thing.